okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.java

```diff
@@ -0,0 +1,291 @@
+import okhttp3.ResponseBody;
+import okhttp3.internal.platform.Platform;
+import okhttp3.internal.publicsuffix.PublicSuffixDatabase;
+import okio.ByteString;
+
+import static okhttp3.internal.Util.addSuppressedIfPossible;
+
+/**
+ * DNS over HTTPS implementation.
+ *
+ * Implementation of https://tools.ietf.org/html/draft-ietf-doh-dns-over-https-11
+ *
+ * <blockquote>A DNS API client encodes a single DNS query into an HTTP request
+ * using either the HTTP GET or POST method and the other requirements
+ * of this section.  The DNS API server defines the URI used by the
+ * request through the use of a URI Template.</blockquote>
+ *
+ * <h3>Warning: This is a non-final API.</h3>
+ *
+ * <p><strong>As of OkHttp 3.11, this feature is an unstable preview: the API is subject to change,
+ * and the implementation is incomplete. We expect that OkHttp 3.12 or 3.13 will finalize this API.
+ * Until then, expect API and behavior changes when you update your OkHttp dependency.</strong>
+ */
+public class DnsOverHttps implements Dns {
+  public static final MediaType DNS_MESSAGE = MediaType.parse("application/dns-message");
+  public static final MediaType UDPWIREFORMAT = MediaType.parse("application/dns-udpwireformat");
+  public static final int MAX_RESPONSE_SIZE = 64 * 1024;
+  private final OkHttpClient client;
+  private final HttpUrl url;
+  private final boolean includeIPv6;
+  private final boolean post;
+  private final MediaType contentType;
+  private final DnsFallbackStrategy dnsFallbackStrategy;
+  private final Dns systemDns;
+
+  DnsOverHttps(Builder builder) {
+    if (builder.client == null) {
+      throw new NullPointerException("client not set");
+    }
+    if (builder.url == null) {
+      throw new NullPointerException("url not set");
+    }
+
+    this.url = builder.url;
+    this.includeIPv6 = builder.includeIPv6;
+    this.post = builder.post;
+    this.contentType = builder.contentType;
+    this.dnsFallbackStrategy = builder.dnsFallbackStrategy;
+    this.client = builder.client.newBuilder().dns(buildBootstrapClient(builder)).build();
+    this.systemDns = builder.systemDns;
+  }
+
+  private static Dns buildBootstrapClient(Builder builder) {
+    List<InetAddress> hosts = builder.bootstrapDnsHosts;
+
+    if (hosts != null) {
+      return new BootstrapDns(builder.url.host(), hosts);
+    } else {
+      return builder.systemDns;
+    }
+  }
+
+  public HttpUrl url() {
+    return url;
+  }
+
+  public boolean post() {
+    return post;
+  }
+
+  public boolean includeIPv6() {
+    return includeIPv6;
+  }
+
+  public MediaType contentType() {
+    return contentType;
+  }
+
+  public OkHttpClient client() {
+    return client;
+  }
+
+  public DnsFallbackStrategy dnsFallbackStrategy() {
+    return dnsFallbackStrategy;
+  }
+
+  public Dns systemDns() {
+    return systemDns;
+  }
+
+  @Override public List<InetAddress> lookup(String hostname) throws UnknownHostException {
+    @Nullable UnknownHostException firstUhe = null;
```

Comment:
nit: we don’t typically use `@Nullable` within method bodies because it’s almost always implied